### PR TITLE
LevelGroup: safely store or filter all characters

### PR DESCRIPTION
**Pull request by breville**
We were encountering issues with a variety of characters submitted as part of a LevelGroup, particularly in the JSON that contained results for all sublevels.  JSON.stringify was escaping them successfully on the client, but we needed to URI encode them to get them to save properly on the server.

It took specific testing to ensure that this round-tripping worked for both the LevelGroup's collective JSON result and the optional UserLevel result that can override this.  This text was used to test:

```
one "two" three
!@#$%^&*()_+-=~`\oh/oh\n\t end\"
donkey
```

We also now filter emoji from the responses, replacing them with small box characters, because our database will not write them.

This change also includes fixes for a couple issues with unavailable level data, exposed by the issues fixed above.

This change also UI tests the various characters saving and restoring properly for embedded free_response and text_match levels, though unfortunately I couldn't work out how to inject emoji characters successfully.

### CODE REVIEW

**Comment by breville** [comment]
fyi @bzotto :)

File: apps/src/code-studio/levels/levelGroup.js [hunk 1]
```diff
@@ -89,6 +89,7 @@ window.initLevelGroup = function (
    *  "1939": {"result": "2,1", "valid": true}}
    */
   function getResult() {
+
```
Comment (mehalshah) [inline apps/src/code-studio/levels/levelGroup.js]: Remove?

File: dashboard/app/helpers/users_helper.rb [hunk 2]
```diff
@@ -132,7 +132,11 @@ def get_pages_completed(user, sl)
     if level.is_a? LevelGroup
       pages_completed = []
 
-      last_attempt = JSON.parse(user.last_attempt(level).level_source.data)
+      if user.last_attempt(level).level_source
```
Comment (mehalshah) [inline dashboard/app/helpers/users_helper.rb:135]: I'd suggest removing the condition. Then on line 158, begin the conditional with last_attempt

**Comment by mehalshah** [comment]
I suggest adding this change to the UI test for free response

https://github.com/code-dot-org/code-dot-org/blob/54c76c540fe3dd146dc81c54057d313d95597d37/dashboard/test/ui/features/free_response_submittable.feature#L26

**Comment by breville** [comment]
Good idea on the test.  It turns out a % in a standalone (i.e. non-LevelGroup embedded) free response doesn't submit successfully.  I don't believe this is a regression since we're not touching that codepath here, but I think it should be tracked separately, since it's probably been that way for a while.

**Comment by mehalshah** [comment]
I'll wait for you to merge this and then start on it

File: apps/src/code-studio/levels/levelGroup.js [hunk 3]
```diff
@@ -144,6 +146,28 @@ window.initLevelGroup = function (
...
+  // Returns the updated string.
+  // Source: http://crocodillon.com/blog/parsing-emoji-unicode-in-javascript
+  function replaceEmoji(source, replace) {
+    /* An issue in Babel is preventing us from using this version.
+    var ranges = [
+      '\ud83c[\udf00-\udfff]', // U+1F300 to U+1F3FF
+      '\ud83d[\udc00-\ude4f]', // U+1F400 to U+1F64F
+      '\ud83d[\ude80-\udeff]'  // U+1F680 to U+1F6FF
+    ];
+    */
```
Comment (mehalshah) [inline apps/src/code-studio/levels/levelGroup.js]: Remove this comment?

**Comment by mehalshah** [comment]
LGTM but please augment existing unit and UI tests

File: apps/src/code-studio/levels/levelGroup.js [hunk 4]
```diff
@@ -10,6 +10,8 @@ window.initLevelGroup = function (
   currentPage,
   lastAttempt) {
 
+  var blankCharacter = "\u25A1";
```
Comment (mehalshah) [inline apps/src/code-studio/levels/levelGroup.js]: You can define this as a constant

File: apps/src/code-studio/levels/levelGroup.js [hunk 5]
```diff
@@ -144,6 +146,21 @@ window.initLevelGroup = function (
     }
   }
 
+  // Replaces emoji in a string with the given string.
+  // Returns the updated string.
+  // Source: http://crocodillon.com/blog/parsing-emoji-unicode-in-javascript
+  function replaceEmoji(source, replace) {
```
Comment (mehalshah) [inline apps/src/code-studio/levels/levelGroup.js]: We're always replacing it with blankCharacter, so drop replace and just sub in blankCharacter

File: dashboard/test/ui/step_definitions/steps.rb [hunk 6]
```diff
@@ -432,6 +441,14 @@ def color_for_status(status)
   element_value_is(selector, expected_value)
 end
 
+Then /^element "([^"]*)" has escaped value "([^"]*)"$/ do |selector, expected_value|
+  element_value_is(selector, YAML.load(%Q(---\n"#{expected_value}"\n)))
+end
+
+Then /^element "([^"]*)" has escaped value '([^']*)'$/ do |selector, expected_value|
+  element_value_is(selector, YAML.load(%Q(---\n"#{expected_value.gsub('"', '\"')}"\n)))
+end
```
Comment (mehalshah) [inline dashboard/test/ui/step_definitions/steps.rb:450]: I feel like this can be combined through regex magic. But its okay to leave it for readability

**Comment by mehalshah** [comment]
Few comments, then ship

**Pull request merged**